server/configHelpers.js

```diff
@@ -0,0 +1,179 @@
+    } else throw e;
+  }
+
+  let configTxt = '';
+  for (let key in clientConfig.options) {
+    const configKey = key
+      .replace('-', '')
+      .replace('_', '')
+      .toLowerCase();
+    const text = `${configKey}: ${clientConfig.options[key]}\n`;
+    configTxt = configTxt.concat(text);
+  }
+  if (!fs.existsSync(clientsPath)) {
+    logger.warn(
+      `Could not find requested client directory at ${clientsPath}. Creating new one...`
+    );
+    fs.mkdirpSync(clientsPath);
+  }
+  fs.writeFileSync(`${clientsPath}/${clientConfig.str('id')}.conf`, configTxt);
+  return clientConfig;
+}
+
+/*
+ * Retrieve a config from clients directory
+ * @param {string} id - id of client to retrieve
+ * @returns {bcfg.Config} - bcfg object of config
+ */
+
+function getConfig(id) {
+  assert(typeof id === 'string', 'Client config must have an id');
+
+  const appConfig = loadConfig('bpanel');
+  const clientsDir = appConfig.str('clients-dir', 'clients');
+
+  const clientsPath = resolve(appConfig.prefix, clientsDir);
+  const config = loadConfig(id, { id, prefix: clientsPath });
+
+  const path = resolve(clientsPath, `${id}.conf`);
+  if (!fs.existsSync(path)) {
+    const error = new Error(`File ${id}.conf not found`);
+    error.code = 'ENOENT';
+    throw error;
+  }
+
+  config.open(`${id}.conf`);
+  return config;
+}
+
+/*
+ * Simple utility that deletes a config
+ * Does not throw errors if client is not found
+ * NOTE: This does not confirm. Deletion is final!
+ * @param {string} id
+ * @returns {bool} - returns true when operation completed successfully
+ */
+
+function deleteConfig(id) {
+  assert(typeof id === 'string', 'Expected to get id of config to delete');
+  const config = getConfig(id);
+  const path = resolve(config.prefix, `${config.str('id')}.conf`);
+  let success = true;
+  try {
+    fs.unlinkSync(path);
+  } catch (e) {
+    logger.error('Problem removing config:', e);
+    success = false;
+  }
+  return success;
+}
+
+/*
+ * create and test clients based on a passed config
+ * @param {Config} clientConfig
+ * @throws {ClientErrors} - throws if at least one client fails
+ * @returns void
+ */
+
+async function testConfigOptions(options) {
+  let clientConfig;
+  try {
+    clientConfig = getConfigFromOptions(options);
+  } catch (e) {
+    throw e;
+  }
+
+  const clients = clientFactory(clientConfig);
+
+  const clientErrors = new ClientErrors();
+
+  // check each client to see if it can connect
+  // for each failure, `addFailed` to the error object
+  for (let key in clients) {
+    // keys come back of the form "nodeClient"
+    // to get the type we need to remove "Client" from the string
+    const type = key.substr(0, key.indexOf('Client'));
+    try {
+      logger.info(`Checking ${key} for config "${clientConfig.str('id')}"`);
+      if (clientConfig.bool(type, true)) await clients[key].getInfo();
+    } catch (e) {
+      clientErrors.addFailed(type, e);
+      continue;
```

Comment:
Will the loop not continue here without the `continue` keyword?

Reply:
I could probably take that out actually. For some reason this formulation was seeming to take longer on failures, so I tried this.